File: cpp/src/Motion/ServoZeroTurn.cpp
//#include "ctre/phoenix/Motion/ServoZeroTurn.h"
//#include "HAL/DriverStation.h"
//#include <math.h>
//
//namespace CTRE {
//namespace Phoenix {
//namespace Motion {
//
//ServoZeroTurn::ServoZeroTurn(CTRE::Phoenix::Drive::ISmartDrivetrain *driveTrain,
//		CTRE::Phoenix::Drive::Styles::Smart smartStyle, float targetHeading,
//		float headingTolerance, ServoParameters *Params, float maxOutput) {
//	_driveTrain = driveTrain;
//	_selectedStyle = smartStyle;
//
//	_targetHeading = targetHeading;
//	_headingTolerance = headingTolerance;
//
//	_maxOutput = maxOutput;
//
//	servoParams->P = Params->P;
//	servoParams->I = Params->I;
//	servoParams->D = Params->D;
//}
//ServoZeroTurn::ServoZeroTurn(CTRE::Phoenix::Drive::ISmartDrivetrain *driveTrain,
//		CTRE::Phoenix::Drive::Styles::Smart smartStyle) {
//	_driveTrain = driveTrain;
//	_selectedStyle = smartStyle;
//}
//bool ServoZeroTurn::Set(float targetHeading, float headingTolerance,
//		float maxOutput) {
//	_maxOutput = maxOutput;
//	return ZeroTurn(targetHeading, headingTolerance);
//}
//float ServoZeroTurn::GetEncoderHeading() {
//	return _driveTrain->GetEncoderHeading();
//}
//void ServoZeroTurn::OnStart() {
//	_isDone = false;
//	_isGood = 0;
//	_state = 0;
//}
//void ServoZeroTurn::OnStop() {
//	_driveTrain->Set(CTRE::Phoenix::Drive::Styles::Smart::PercentOutput, 0, 0);
//	_isDone = true;
//}
//bool ServoZeroTurn::IsDone() {
//	return _isDone;
//}
//void ServoZeroTurn::OnLoop() {
//	switch (_state) {
//	case 0:
//		_driveTrain->SetPosition(0.0f);
//		_state = 1;
//		break;
//	case 1:
//		bool running = ZeroTurn(_targetHeading, _headingTolerance);
//
//		if (running == true)
//			_isGood = 0;
//		else if (_isGood < 10)
//			++_isGood;
//		else {
//			_driveTrain->Set(CTRE::Phoenix::Drive::Styles::Smart::PercentOutput, 0, 0);
//			_isDone = true;
//		}
//		break;
//	}
//}
//bool ServoZeroTurn::ZeroTurn(float targetHeading, float headingTolerance) {
//	if (servoParams->P == 0 && servoParams->I == 0 && servoParams->D == 0)
//		HAL_SendError(true, 1, false,
//				"CTR: Servo Zero Turn has no PID values, cannot turn", "", "",
//				true);
//	/* Grab the current heading*/
//	float currentHeading = GetEncoderHeading();
//
//	/* Find the difference between last heading and current heading */
//	_timeElapsed = _myStopwatch->Duration();
//	float headingRate = currentHeading - _previousHeading;
//	_myStopwatch->Start();
//
//	/* Heading PID */
//	float headingError = targetHeading - currentHeading;
//	float X = ((headingError) * servoParams->P)
//			- ((headingRate) * servoParams->D);
//	X = -X;
//	X = fmax(-_maxOutput, fmin(X, _maxOutput));
//
//	/** Set the output of the drivetrain */
//	/** Set the output of the drivetrain */
//	switch (_selectedStyle) {
//	case CTRE::Phoenix::Drive::Styles::Smart::PercentOutput:
//		_driveTrain->Set(CTRE::Phoenix::Drive::Styles::Smart::PercentOutput, 0, X);
//		break;
//	case CTRE::Phoenix::Drive::Styles::Smart::Voltage:
//		_driveTrain->Set(CTRE::Phoenix::Drive::Styles::Smart::Voltage, 0, X);
//		break;
//	case CTRE::Phoenix::Drive::Styles::Smart::VelocityClosedLoop:
//		_driveTrain->Set(CTRE::Phoenix::Drive::Styles::Smart::VelocityClosedLoop, 0, X);
//		break;
//	}
//
//	/** Grab the heading to compare next time */
//	_previousHeading = currentHeading;
//
//	if (fabs(headingError) >= headingTolerance) {
//		_isRunning = true;
//	} else {
//		_isRunning = false;
//	}
//	return _isRunning;
//}
//
//}
//}
//}

File: java/src/com/ctre/phoenix/motion/ServoZeroTurn.java
package com.ctre.phoenix.motion;
//package com.ctre.phoenix.Motion;
//
//import com.ctre.phoenix.Util;
//import com.ctre.phoenix.Drive.*;
//import com.ctre.phoenix.motorcontrol.*;
//import com.ctre.phoenix.Mechanical.*;
//import com.ctre.phoenix.Time.*;
//
//public class ServoZeroTurn implements com.ctre.phoenix.ILoopable
//{
//	ISmartDrivetrain _driveTrain;
//    Styles.Smart _selectedStyle;
//	
//    private ServoParameters servoParams = new ServoParameters();
//	public ServoParameters GetServoParams() { return servoParams; }
//	public void SetServoParams(ServoParameters params)
//	{
//		servoParams = params;
//	}
//	
//    StopWatch _myStopwatch = new StopWatch();
//
//    double _targetHeading;
//    double _previousHeading;
//    double _timeElapsed;
//
//    boolean _isRunning = false;
//    boolean _isDone = false;
//    byte _isGood = 0;
//
//    /** Constructor */
//    public ServoZeroTurn(ISmartDrivetrain driveTrain, Styles.Smart smartStyle, double targetHeading, ServoParameters Params)
//    {
//        _driveTrain = driveTrain;
//        _selectedStyle = smartStyle;
//
//        _targetHeading = targetHeading;
//
//        SetServoParams(Params);
//    }
//
//    public ServoZeroTurn(ISmartDrivetrain driveTrain, Styles.Smart style)
//    {
//        _driveTrain = driveTrain;
//        _selectedStyle = style;
//    }
//
//    public void set(double targetHeading)
//    {
//		_targetHeading = targetHeading;
//    }
//
//    public double getEncoderHeading()
//    {
//        return _driveTrain.GetEncoderHeading();
//    }
//
//    private void ZeroTurn(double targetHeading)
//    {
//    	if(servoParams.P == 0 && servoParams.I == 0 && servoParams.D == 0)
//    		com.ctre.phoenix.CTRLogger.Log(-503, "Servo Zero Turn");
//        /* Grab the current heading*/
//        double currentHeading = GetEncoderHeading();
//
//        /* Find the difference between last heading and current heading */
//        _timeElapsed = _myStopwatch.getDuration();
//        double headingRate = currentHeading - _previousHeading;
//        _myStopwatch.start();
//
//        /* Heading PID */
//        double headingError = targetHeading - currentHeading;
//        double X = servoParams.PID(headingError, headingRate);
//        X = -X;
//
//
//        /** Set the output of the drivetrain */
//        /** Set the output of the drivetrain */
//        switch (_selectedStyle)
//        {
//            case PercentOutput:
//                _driveTrain.set(Styles.Smart.PercentOutput, 0, X);
//                break;
//            case Voltage:
//                _driveTrain.set(Styles.Smart.Voltage, 0, X);
//                break;
//            case VelocityClosedLoop:
//                _driveTrain.set(Styles.Smart.VelocityClosedLoop, 0, X);
//                break;
//        }
//
//        /** Grab the heading to compare next time */
//        _previousHeading = currentHeading;
//    }
//
//    /** ILoopable */
//    public void onStart()
//    {
//        _isDone = false;
//        _isGood = 0;
//		servoParams.onStart();
//    }
//
//    public void onStop()
//    {
//        _driveTrain.set(Styles.Basic.PercentOutput, 0, 0);
//        _isDone = true;
//    }
//
//    public boolean isDone()
//    {
//        return _isDone;
//    }
//
//    public void onLoop()
//    {
//		if(!_isDone)
//		{
//			ZeroTurn(_targetHeading);
//
//			if (servoParams.isDone())
//			{
//				_driveTrain.set(Styles.Basic.PercentOutput, 0, 0);
//				_isDone = true;
//			}
//        }
//    }
//}